grpc/src/main/java/com/linecorp/armeria/internal/client/grpc/ArmeriaClientCall.java

```diff
@@ -491,37 +543,34 @@ private void close(Status status, Metadata metadata) {
                                                MILLISECONDS.toNanos(ctx.responseTimeoutMillis()) + "ns.");
         }
 
-        final RequestLogBuilder logBuilder = ctx.logBuilder();
         final StatusAndMetadata statusAndMetadata = new StatusAndMetadata(status, metadata);
+
+        // We want to close the user-facing ClientCall.Listener right away, before we move on to recycling
+        // underlying resources.
+        executor.execute(() -> {
+            try (SafeCloseable ignored = ctx.push()) {
+                assert listener != null;
+                listener.onClose(statusAndMetadata.status(), statusAndMetadata.metadata());
+            } catch (Throwable t) {
+                closeWhenListenerThrows(t);
```

Comment:
`closeWhenListenerThrows(t)` will perform nothing because `closed` is true already.
Should we just leave a warning log as a last resort?

```java
logger.warn("Unexpected exception while closing {}", listener, t);
```

Reply:
It's a good point. Thanks!